### ISSUE TITLE
Parent.id is set to a non-Elastic APM Span/Transaction

### ISSUE THREAD

**Issue by gregkalapos** [elastic/apm-agent-dotnet#883]
Found in Opbeans and also [here](https://discuss.elastic.co/t/nothing-to-see-in-traces/236917).

One example of such an id: `|eb88cbba-4040e4007bc3c335.`.

My quick slack analysis:

```
that id looks like an id coming from Activity https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.activity?view=netcore-3.1
Activity is a “tracing library” build into .NET (fun fact the C# OpenTelemetry SIG decided  to take this API and turn it into an  OpenTelemetry compatible API - so think it’s like a span) -  anyways, so what we do in the agent is that we try to keep our trace id in sync with  the traceid on Activity (that way if users look at Activity.Current.TraceId then it’ll be the same as our trace id in elastic). Problem is Activity pre-dates W3C TraceContext and - originally it had a hierarchical id format (that’s what you pasted), but now it also supports w3c format. We only keep the id in sync when  it’s w3c format: https://github.com/elastic/apm-agent-dotnet/blob/master/src/Elastic.Apm/Model/Transaction.cs#L85
At least that’s my intention. I can see 2 options here: Either I messed up and somewhere a check is missing or it’s in W3C format, but it’s parent is not, so the check is ok, but still we set the parent id which is in the old format - not sure it’s possible.
I’ll dig deeper - I still just start my day here. But thanks for mentioning this! that id looks like an id coming from Activity https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.activity?view=netcore-3.1
Activity is a “tracing library” build into .NET (fun fact the C# OpenTelemetry SIG decided  to take this API and turn it into an  OpenTelemetry compatible API - so think it’s like a span) -  anyways, so what we do in the agent is that we try to keep our trace id in sync with  the traceid on Activity (that way if users look at Activity.Current.TraceId then it’ll be the same as our trace id in elastic). Problem is Activity pre-dates W3C TraceContext and - originally it had a hierarchical id format (that’s what you pasted), but now it also supports w3c format. We only keep the id in sync when  it’s w3c format: https://github.com/elastic/apm-agent-dotnet/blob/master/src/Elastic.Apm/Model/Transaction.cs#L85
At least that’s my intention. I can see 2 options here: Either I messed up and somewhere a check is missing or it’s in W3C format, but it’s parent is not, so the check is ok, but still we set the parent id which is in the old format - not sure it’s possible.
I’ll dig deeper - I still just start my day here. But thanks for mentioning this! ```

**Comment by arindamdat** [comment]
I am facing an issue probably related to this, where I see the "Traces" tab on APM UI empty.
More details can be found in this Slack conversation with @dgieselaar **[here](https://elasticstack.slack.com/archives/CNKF2LEMP/p1592409333054200)**.

**Comment by gregkalapos** [comment]
I reproduced this and unfortunately as it seems the following combination is also possible:

`Activity.Current.IdFormat == ActivityIdFormat.W3C` BUT `Activity.Current.ParentId` is in the old hierarchical format; yet we set  `Activity.Current.ParentId` as parent id; so that's already a problem.

<img width="815" alt="image" src="https://user-images.githubusercontent.com/1091853/85744717-64acff00-b705-11ea-86de-b65c464f571f.png">

But the bigger issue here is indeed the fact that ASP.NET Core generates the tracecontext header out of the box, even if there is no APM tool involved. So if we have 2 services: A -> B (A calls B) and we only have Elastic APM in B, then we'll read the traceparent header from A with an id that points to a parent.id which is outside elastic APM. 

This issue can also happen if A is monitored by another tool and B is monitored by Elastic - but in ASP.NET Core 3+ we don't even need another tool to have this problem.

**Comment by arindamdat** [comment]
I see the following:
ASP.Net core 3.1
Service A:
Activity.Current:
![image](https://user-images.githubusercontent.com/7764767/85765583-e025a500-b733-11ea-9541-0cfc485cdf9c.png)
Service-B:
Request Headers:
![image](https://user-images.githubusercontent.com/7764767/85765624-e9af0d00-b733-11ea-96a6-e80938c9e212.png)
Activity.Current:
![image](https://user-images.githubusercontent.com/7764767/85765674-f29fde80-b733-11ea-9c09-230992983e40.png)

**Comment by arindamdat** [comment]
And using Elastic APM agent at both service end.

**Comment by gregkalapos** [comment]
Thanks @arindamdat! That scenario is also reproducible - in that case, already in ServiceA we have an issue, which is that we set the `ParentId` of the transaction in ServiceA to the activity parent id which was created by ASP.NET Core - and that is why there won't by any transaction there where the `ParentId` is `null`. And a trace must begin with a transaction where the `ParentId` is null, so that is why you don't see traces. 


The [PR I just opened](https://github.com/elastic/apm-agent-dotnet/pull/888) also covers that issue.

**Comment by gregkalapos** [comment]
Fixed in https://github.com/elastic/apm-agent-dotnet/pull/888 - will be included in the next release.

**Comment by arindamdat** [comment]
Hi @gregkalapos , is this fix part of ElasticApm .net agent 1.6.1?

**Comment by gregkalapos** [comment]
Hi @arindamdat, yes it is. It was released in version `1.6.0`.
